pkg/controllers/subnet/subnet_controller.go

```diff
@@ -70,8 +73,41 @@ func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		log.Error(err, "Unable to fetch Subnet CR", "req", req.NamespacedName)
 		return ResultRequeue, err
 	}
+
+	bindingCRs := r.getSubnetBindingCRsBySubnet(ctx, subnetCR)
+	if len(bindingCRs) > 0 {
+		if !controllerutil.ContainsFinalizer(subnetCR, servicecommon.SubnetFinalizerName) {
+			controllerutil.AddFinalizer(subnetCR, servicecommon.SubnetFinalizerName)
+			if err := r.Client.Update(ctx, subnetCR); err != nil {
+				log.Error(err, "Failed to add the finalizer", "Subnet", req.NamespacedName)
+				msgFailAddFinalizer := fmt.Sprintf("Failed to add the finalizer on a Subnet for the reference by SubnetConnectionBindingMap %s", bindingCRs[0].Name)
+				r.StatusUpdater.UpdateFail(ctx, subnetCR, err, "Failed to add the finalizer on Subnet used by SubnetConnectionBindingMaps", setSubnetReadyStatusFalse, msgFailAddFinalizer)
+				return ResultRequeue, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(subnetCR, servicecommon.SubnetFinalizerName) {
+			controllerutil.RemoveFinalizer(subnetCR, servicecommon.SubnetFinalizerName)
+			if err := r.Client.Update(ctx, subnetCR); err != nil {
+				log.Error(err, "Failed to delete the finalizer", "Subnet", req.NamespacedName)
+				msgFailDelFinalizer := "Failed to remove the finalizer on a Subnet when there is no reference by SubnetConnectionBindingMaps"
+				r.StatusUpdater.UpdateFail(ctx, subnetCR, err, "Failed to delete the finalizer from Subnet", setSubnetReadyStatusFalse, msgFailDelFinalizer)
+				return ResultRequeue, err
+			}
+		}
+	}
+
 	if !subnetCR.DeletionTimestamp.IsZero() {
 		r.StatusUpdater.IncreaseDeleteTotal()
+		bindingsOnNSX := r.getNSXSubnetBindingsBySubnet(string(subnetCR.UID))
+		if len(bindingsOnNSX) > 0 {
+			err := fmt.Errorf("failed to delete Subnet CR %s", req.String())
```

Comment:
do we need to print the whole req? can just use req.NamespacedName?

Reply:
The implementation of `req.String` would print "$namespac/$name", it is equal to req.NamespacedName. This is because the only property in Request is types.NamespacedName.